### PR TITLE
fix(plugins): rate limiter returns proper HTTP status codes and headers (#2668)

### LINKED ISSUE #2668: [BUG]: Follow-up: Rate limiter — incorrect HTTP status, missing headers, and feature gaps (refs #2397)
### Summary

Manual testing of the **RateLimiterPlugin** shows that rate-limiting metadata is returned in the JSON response but **not exposed via standard HTTP headers**, and the **HTTP status remains 200 instead of 429** when a request is limited. In addition, several expected features are missing (IP-based limiting, Redis backend, daily limits, MCP/tool coverage).

This issue is a follow-up to **#2397** and consolidates reproduction steps, logs, observed vs expected behavior, root cause analysis, and suggested fixes.

---

## Background

The `RateLimiterPlugin` correctly computes and returns rate-limit metadata, but the **RPC/HTTP layer does not translate that metadata into HTTP response semantics** (headers + status code).

Relevant code paths:

* `rate_limiter.py` – plugin returns rate-limit metadata
* `main.py` – RPC/HTTP handler does not map metadata to headers or status

---

## Environment

* **Repository / Branch / Tag / Commit**:
  `main` / `v1.0.0-BETA-2` / `a44ca2b5b2b541dc5315475b78d9be9de1afde48`
* **Date of test**: 2026-01-29
* **OS**: Linux (development)
* **Gateway**: `GUNICORN_WORKERS=1`
* **Plugins enabled**: `PLUGINS_ENABLED=true`
* **Plugin under test**: `RateLimiterPlugin` (enforce mode)
* **Config**: `config.yaml`

---

## Test Summary

* **Test cases executed**: 12
* **Passing**: 7
* **Partial**: 3

  * Headers missing
  * Daily unit (`/d`) not supported
  * Plugin returns error JSON but HTTP status is 200
* **Failing**: 2

  * IP-based limiting
  * Redis / distributed limiting

---

## Reproduction Steps

1. Start the gateway with `config.yaml`
   (e.g. `make dev` or `python -m mcpgateway.main`)
2. Repeatedly call a REST tool until rate limits are exceeded
3. Observe:

   * HTTP status
   * Response headers
   * JSON response body

---

## Observed Behavior

* **HTTP status**: `200 OK` returned even when the rate limit is exceeded
  (expected: `429 Too Many Requests`)
* **HTTP headers**:
  No `X-RateLimit-*` or `Retry-After` headers present
* **JSON body**:
  Rate-limit metadata is present under `error.data.details`, e.g.
* **Scope limitation**:
  Enforcement applies only to REST tools; **MCP tools/endpoints are not covered**
* **Other gaps**:

  * IP-based limiting for unauthenticated requests not implemented
  * Daily unit (`/d`) not supported by `_parse_rate()`
  * No Redis backend for distributed enforcement

---

## Expected Behavior

When a request is rate-limited:

* **HTTP status**: `429 Too Many Requests`
* **Standard headers should be present**:

  * `X-RateLimit-Limit`
  * `X-RateLimit-Remaining`
  * `X-RateLimit-Reset` (unix epoch)
  * `Retry-After`
* Rate-limit metadata may still be included in the JSON body, but **HTTP semantics must be honored**
* Rate limiting should be enforced consistently across:

  * REST
  * RPC
  * MCP tools/endpoints
* Supported units should include `/s`, `/m`, `/h`, and `/d`
* Distributed deployments should be supported (e.g. Redis backend)

---

## Root Cause

The plugin computes and returns rate-limit metadata, but the **RPC/HTTP handler does not**:

* Map that metadata into HTTP response headers
* Set an appropriate HTTP status code (`429`)

Additionally, several features (IP-based limiting, Redis backend, daily unit support) are not implemented in the plugin.

---

## Impact

* Clients, SDKs, and proxies relying on **standard HTTP rate-limit semantics** cannot react correctly
* Automated retry logic, observability, and monitoring are impaired
* In-memory-only storage prevents correct enforcement in multi-process or multi-host deployments

---

### High Priority

1. **Return HTTP 429** for rate-limited requests
2. **Add standard rate-limit headers**:

   * `X-RateLimit-Limit`
   * `X-RateLimit-Remaining`
   * `X-RateLimit-Reset`
   * `Retry-After`
3. **Extend enforcement to all relevant endpoints**, including MCP tools (not just REST tools)

### Medium Priority

4. Implement **IP-based limiting** for anonymous/unauthenticated requests
5. Support **daily unit (`/d`)** in `_parse_rate()` and config parsing

---

## Future Enhancements

6. Per-endpoint rate-limit configuration (e.g. `endpoints:` block in `config.yaml`)
7. Redis (or other) backend for distributed deployments
8. Align header semantics with RFCs (e.g. RFC 6585) and consider adding:

   * Config validation endpoint or CLI
   * Hot-reload safety tests when `RELOAD=true`

---

## Suggested Implementation Notes

* Translate plugin metadata into HTTP response headers in the RPC/HTTP layer (`main.py`)
* When a plugin signals a rate-limit violation, the handler should:

  * Set HTTP status to `429`
  * Add `X-RateLimit-*` and `Retry-After` headers derived from plugin metadata
  * Still return a structured JSON error body with `plugin_name` and `plugin_error_code`
* Add unit and integration tests for:

  * Status code behavior
  * Header presence
  * Distributed (Redis) behavior
* Ensure config merge/reload semantics are idempotent and tested

---

## References

* Original issue: #2397

### PR DESCRIPTION
**Pull request by crivetimihai**
> **Note:** This PR was re-created from #3125 due to repository maintenance. Your code and branch are intact. @ja8zyjits please verify everything looks good.

# 🐛 Bug-fix PR

## 📌 Summary

Fixes #2668
Rate-limiting and other policy enforcement plugins needed the ability to return proper HTTP status codes (e.g., 429 Too Many Requests) and standard headers (e.g., Retry-After, X-RateLimit-*) when violations occur. Previously, all plugin violations returned HTTP 200 with JSON-RPC error payloads, making it impossible for:

   - Rate-limiting plugins to signal backoff periods to clients
   - Authorization plugins to return proper 403 Forbidden responses
   - Content moderation to return 422 Unprocessable Entity
   - Clients to implement proper retry logic based on HTTP semantics

This created a mismatch between HTTP-native clients expecting standard status codes and the gateway's JSON-RPC-centric error handling.

---

## 🔁 Reproduction Steps

1. Start the gateway with config.yaml
(e.g. make dev or python -m mcpgateway.main)

2. Repeatedly call a REST tool until rate limits are exceeded

3. Observe:
    - HTTP status
    - Response headers
    - JSON response body

## 🐞 Root Cause
Improper error code handling in plugin violation.

## 💡 Fix Description

This PR introduces a three-tier approach for determining HTTP status codes in plugin violation responses:

   1. Explicit Status (highest priority): Plugins can set http_status_code directly
   2. Code Mapping (fallback): Common violation codes map to appropriate HTTP statuses via PLUGIN_VIOLATION_CODE_MAPPING
   3. Default 200 (ultimate fallback): Maintains JSON-RPC compliance when no status is specified

#### 1. Enhanced PluginViolation Model
   ```python
   class PluginViolation(BaseModel):
       # ... existing fields ...
       http_status_code: Optional[int] = None  # NEW: Explicit HTTP status
       http_headers: Optional[dict[str, str]] = None  # NEW: Custom headers (e.g., Retry-After)
   ```

#### 2. Violation Code Mapping (mcpgateway/plugins/framework/constants.py)
   ```python
   PLUGIN_VIOLATION_CODE_MAPPING = {
       "RATE_LIMIT": 429,              # Rate limiting
       "INVALID_URI": 400,             # Bad request
       "PROTOCOL_BLOCKED": 403,        # Forbidden
       "CONTENT_TOO_LARGE": 413,       # Payload too large
       "CONTENT_MODERATION": 422,      # Unprocessable entity
       "PII_DETECTED": 422,            # Sensitive data
       "INVALID_TOKEN": 401,           # Unauthorized
       # ... 20+ mappings total
   }
   ```

 #### 3. Updated Exception Handler (mcpgateway/main.py)
   ```python
   async def plugin_violation_exception_handler(_request: Request, exc: PluginViolationError):
       # Determine HTTP status: explicit → mapping → default 200
       http_status = exc.violation.http_status_code if exc.violation and exc.violation.http_status_code else None
       if not http_status:
           http_status = PLUGIN_VIOLATION_CODE_MAPPING.get(exc.violation.code, 200)

       # Add custom headers if provided
       headers = exc.violation.http_headers if exc.violation and exc.violation.http_headers else None
       response = ORJSONResponse(status_code=http_status, content={...})
       if headers:
           response.headers.update(headers)
       return response
   ```

   ---

   ## 🧪 Testing

   Added **8 comprehensive test cases** covering:

   ✅ Explicit http_status_code takes precedence over mapping
   ✅ Code mapping fallback (e.g., RATE_LIMIT → 429)
   ✅ Default to 200 for unknown codes (JSON-RPC compliance)
   ✅ HTTP headers properly propagated to response
   ✅ Multiple headers (Retry-After, X-RateLimit-*) included correctly
   ✅ Graceful handling when headers is None
   ✅ Backward compatibility with existing violations (no fields set)
   ✅ Doctest examples updated to reflect new behavior

   **Test Coverage:** All new code paths covered, including edge cases.

   ---
## End to End test

1. In `.env` file update:
```
PLUGINS_ENABLED=true
RATE_LIMITER_ENABLED=true
PLUGINS_CONFIG_FILE=plugins/config.yaml
DEV_MODE=true
```
2. In `plugins/config.yaml` update:
```diff
--- a/plugins/config.yaml
+++ b/plugins/config.yaml
@@ -208,20 +208,21 @@ plugins:

   # Rate limiter (fixed window, in-memory)
   - name: "RateLimiterPlugin"
+    enabled: true
     kind: "plugins.rate_limiter.rate_limiter.RateLimiterPlugin"
     description: "Per-user/tenant/tool rate limits"
     version: "0.1.0"
     author: "Mihai Criveti"
     hooks: ["prompt_pre_fetch", "tool_pre_invoke"]
     tags: ["limits", "throttle"]
-    mode: "disabled"
-    priority: 20
+    mode: "enforce"
+    priority: 5
     conditions: []
     config:
-      by_user: "60/m"
-      by_tenant: "600/m"
+      by_user: "6/m"
+      by_tenant: "6/m"
       by_tool:
-        search: "10/m"
+        search: "1/m"
```
3. Also update your environment with the following changes:
```bash
$ export RATE_LIMITER_ENABLED=true
$ export RATE_LIMIT_DEFAULT=5/m
$ export LOG_LEVEL=DEBUG
$ export SERVER_URL=http://localhost:8000
```
4. Create a JWT token and assign it to your `TOKEN` environment variable
```bash
export TOKEN="$(                                                                                                                                                                        ─╯
    python -m mcpgateway.utils.create_jwt_token \
        --username admin@example.com \
        --exp 10080 \
        --secret "$JWT_SECRET_KEY" \
        --algo HS256
)"
```
5. Start the mcp gateway with `make dev`
6. Once the gateway is ready, create a tool (this dummy tool does a rest call to an external api)
```bash
curl -X 'POST' \
  'http://localhost:8000/tools/' \
  -H 'accept: application/json' \
  -H "Authorization: Bearer $TOKEN" \
  -H 'Content-Type: application/json' \
  -d '{
  "tool": {
    "allow_auto": false,
    "name": "dummy",
    "displayName": "Dummy API",
    "url": "https://fakerestapi.azurewebsites.net/api/v1/Users/1",
    "description": "Dummmy API for testing",
    "integration_type": "REST",
    "request_type": "GET",
    "headers": {
      "accept": "*/*"
    },
    "inputSchema": {},
    "outputSchema": {},
    "annotations": {},
    "jsonpath_filter": "",
    "visibility": "public",
    "expose_passthrough": true
  }
}'
```
7. Call the tool repeatedly and check the response
```bash
curl -i -H "Authorization: Bearer $TOKEN" \                                                                                                                                             ─╯
    -H "Content-Type: application/json" \
    -d '{"jsonrpc":"2.0","id":2,"method":"tools/call",
         "params":{"name":"dummy","arguments":{"num1":5,"num2":3}}}' \
    http://localhost:8000/rpc
```
8. After 6 requests within a minute, you can see the the `deny 422` responses with all the required headers
```bash
HTTP/1.1 429 Too Many Requests
date: Wed, 25 Feb 2026 22:20:33 GMT
server: uvicorn
content-length: 489
content-type: application/json
x-ratelimit-limit: 6
x-ratelimit-remaining: 0
x-ratelimit-reset: 1772058092
retry-after: 59
x-content-type-options: nosniff
x-frame-options: DENY
x-xss-protection: 0
x-download-options: noopen
referrer-policy: strict-origin-when-cross-origin
content-security-policy: default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval' https://cdnjs.cloudflare.com https://cdn.tailwindcss.com https://cdn.jsdelivr.net https://unpkg.com; style-src 'self' 'unsafe-inline' https://cdnjs.cloudflare.com https://cdn.jsdelivr.net; img-src 'self' data: https:; font-src 'self' data: https://cdnjs.cloudflare.com; connect-src 'self' ws: wss: https:; frame-ancestors 'none';
x-correlation-id: 0b9ff1098fbf4f11b4d68afc133d7926

{"error":{"code":-32602,"message":"Plugin Violation: Rate limit exceeded for tool dummy, user admin@example.com, or tenant default","data":{"description":"Rate limit exceeded for tool dummy, user admin@example.com, or tenant default","details":{"limited":true,"remaining":0,"reset_in":59,"dimensions":{"violated":[{"limited":true,"remaining":0,"reset_in":59},{"limited":true,"remaining":0,"reset_in":59}],"allowed":[]}},"plugin_error_code":"RATE_LIMIT","plugin_name":"RateLimiterPlugin"}}}
```

   ---

   ## 🔄 Backward Compatibility

   **✅ Fully Backward Compatible**

   - New fields are **optional** (Optional[int], Optional[dict[str, str]])
   - Existing plugins without these fields continue to work (default to HTTP 200)
   - JSON-RPC error structure unchanged
   - No breaking changes to public APIs

   ---

   ## 📝 Example Usage

   ### Rate-Limiting Plugin
   ```python
   return PluginViolation(
       reason="Rate limit exceeded",
       description="Too many requests from this client",
       code="RATE_LIMIT",
       http_status_code=429,  # Explicit status
       http_headers={
           "Retry-After": "60",
           "X-RateLimit-Limit": "100",
           "X-RateLimit-Remaining": "0",
           "X-RateLimit-Reset": "1737394800"
       }
   )
   ```

   **Response:**
   ```http
   HTTP/1.1 429 Too Many Requests
   Retry-After: 60
   X-RateLimit-Limit: 100
   X-RateLimit-Remaining: 0
   X-RateLimit-Reset: 1737394800

   {
     "error": {
       "code": -32602,
       "message": "Plugin Violation: Too many requests from this client",
       "data": {
         "description": "Too many requests from this client",
         "plugin_error_code": "RATE_LIMIT",
         ...
       }
     }
   }
   ```

---

## 🧪 Verification

| Check                                 | Command              | Status |
|---------------------------------------|----------------------|--------|
| Lint suite                            | `make lint`          |    ✅    |
| Unit tests                            | `make test`          |     ✅   |
| Coverage ≥ 80 %                       | `make coverage`      |     ✅   |
| Manual regression no longer fails     | steps / screenshots  |    ✅    |

## 📐 MCP Compliance (if relevant)
- [x] Matches current MCP spec
- [x] No breaking change to MCP clients

## ✅ Checklist
- [x] Code formatted (`make black isort pre-commit`)
- [x] No secrets/credentials committed
- [x] Code follows project style guidelines
- [x] Tests added for new functionality
- [x] Docstrings updated
- [x] Backward compatible
- [x] No breaking changes
- [x] Related issue linked (#2668)
- [x] Conventional commit message format
- [x] DCO sign-off included

---

This PR enables plugins to implement proper HTTP semantics for policy enforcement while maintaining full backward compatibility with the existing JSON-RPC error model.

### CODE REVIEW

File: mcpgateway/plugins/framework/constants.py [hunk 1]
```diff
@@ -46,3 +46,30 @@
 GET_PLUGIN_CONFIG = "get_plugin_config"
 HOOK_TYPE = "hook_type"
 INVOKE_HOOK = "invoke_hook"
+
+# Plugin Violation Code to HTTP Status Code Mapping
+# Maps violation codes to their corresponding HTTP status codes for proper error responses
+PLUGIN_VIOLATION_CODE_MAPPING = {
```
Comment (msureshkumar88) [inline mcpgateway/plugins/framework/constants.py]: Use Enum for clarity + safety
from enum import Enum, unique

@unique
class ViolationCode(Enum):
    # Rate Limiting
    RATE_LIMIT = 429

    # Resource & URI Validation
    INVALID_URI = 400
    PROTOCOL_BLOCKED = 403
    DOMAIN_BLOCKED = 403
    CONTENT_TOO_LARGE = 413

    # Content Moderation & Safety
    CONTENT_MODERATION = 422
    MODERATION_ERROR = 503
    PII_DETECTED = 422
    SENSITIVE_CONTENT = 422

    # Authentication & Authorization
    INVALID_TOKEN = 401
    API_KEY_REVOKED = 401
    AUTH_REQUIRED = 401

    # Generic Violations
    PROHIBITED_CONTENT = 422
    BLOCKED_CONTENT = 403
    BLOCKED = 403
    EXECUTION_ERROR = 500
    PROCESSING_ERROR = 500

File: tests/unit/mcpgateway/test_main.py [hunk 2]
```diff
@@ -3025,7 +3029,8 @@ def test_plugin_violation_exception_handler_with_full_violation(self):
 
         result = asyncio.run(plugin_violation_exception_handler(None, exc))
 
-        assert result.status_code == 200
+        # Verify code mapping works (PROHIBITED_CONTENT -> 422 in PLUGIN_VIOLATION_CODE_MAPPING)
+        assert result.status_code == 422  # Uses mapping
```
Comment (msureshkumar88) [inline tests/unit/mcpgateway/test_main.py]: Better to reuse response codes(422) from constraints.py

File: mcpgateway/main.py [hunk 3]
```diff
@@ -1460,7 +1463,23 @@ async def plugin_violation_exception_handler(_request: Request, exc: PluginViola
...
     json_rpc_error = PydanticJSONRPCError(code=status_code, message="Plugin Violation: " + message, data=violation_details)
-    return ORJSONResponse(status_code=200, content={"error": json_rpc_error.model_dump()})
+
+    # Use HTTP status code from violation if present (e.g., 429 for rate limiting)
+    http_status = exc.violation.http_status_code if exc.violation and exc.violation.http_status_code else None
+    if http_status and not 100 <= http_status <= 599:
+        logger.warning(f"Invalid HTTP status code {http_status} from violation, defaulting to 200")
+        http_status = None
+    if not http_status:
+        logger.info("Using Plugin violation code mapping for lack of http_status_code")
```
Comment (gandhipratik203) [inline mcpgateway/main.py]: would it make sense to consider changing logger.info("Using Plugin violation code mapping...") at mcpgateway/main.py:1473 to debug? This seems like a normal
  request-path fallback (likely to occur frequently when only code is set), so info may add avoidable log noise; debug would be more consistent, while keeping the
  invalid-status log at warning for actual anomalies.

File: mcpgateway/main.py [hunk 4]
```diff
@@ -1460,7 +1463,23 @@ async def plugin_violation_exception_handler(_request: Request, exc: PluginViola
         if exc.violation.plugin_name:
             violation_details["plugin_name"] = exc.violation.plugin_name
     json_rpc_error = PydanticJSONRPCError(code=status_code, message="Plugin Violation: " + message, data=violation_details)
-    return ORJSONResponse(status_code=200, content={"error": json_rpc_error.model_dump()})
+
+    # Use HTTP status code from violation if present (e.g., 429 for rate limiting)
+    http_status = exc.violation.http_status_code if exc.violation and exc.violation.http_status_code else None
+    if http_status and not 100 <= http_status <= 599:
```
Comment (gandhipratik203) [inline mcpgateway/main.py]: Would it be good to tighten the accepted http_status_code range in plugin_violation_exception_handler from 100..599 to 400..599 (mcpgateway/main.py:1469)?
  Because this path always returns an error JSON body, allowing 1xx can lead to non-standard behavior (informational responses are not typically final error responses
  and may be handled inconsistently by clients/proxies). Restricting to 400..599 would make the contract clearer: plugin violations always map to error-class HTTP
  responses, which is generally safer and more predictable for downstream tooling (SDKs, gateways, load balancers, retries, monitoring).

File: mcpgateway/plugins/framework/constants.py (hunk 1 above)
Comment (ja8zyjits) [reply mcpgateway/plugins/framework/constants.py]: I have updated it to frozen dataclass, since Enum wont allow me to have different name and same value.
eg: In Enum, I cannot have `CONTENT_MODERATION = 422` and `PII_DETECTED = 422`. Enum will collapse to `CONTENT_MODERATION = 422`. So when I do `ViolationCode.PII_DETECTED.name` I will get `CONTENT_MODERATION`, instead of `PII_DETECTED`
So it will create aliases

File: mcpgateway/main.py (hunk 4 above)
Comment (ja8zyjits) [reply mcpgateway/main.py]: Done
THanks for the suggestion.

File: mcpgateway/main.py (hunk 3 above)
Comment (ja8zyjits) [reply mcpgateway/main.py]: Done, thanks for the suggestion.

File: mcpgateway/plugins/framework/constants.py (hunk 1 above)
Comment (ja8zyjits) [reply mcpgateway/plugins/framework/constants.py]: Thanks for the suggestion.

File: tests/unit/mcpgateway/test_main.py (hunk 2 above)
Comment (ja8zyjits) [reply tests/unit/mcpgateway/test_main.py]: thanks.
Done

**Comment by ja8zyjits** [comment]
I have updated the `end to end testing` section in the PR description for ease of testing.

Thanks

File: mcpgateway/plugins/framework/constants.py [hunk 5]
```diff
@@ -43,3 +47,45 @@
...
 HOOK_TYPE = "hook_type"
 INVOKE_HOOK = "invoke_hook"
+
+
+@dataclass(frozen=True)
+class PluginViolationCode:
+    """
+    Plugin violation codes as an immutable dataclass object.
+
+    Provide Maping for violation codes to their corresponding HTTP status codes for proper error responses.
```
Comment (gcgoncalves) [inline mcpgateway/plugins/framework/constants.py]: Typo: Mapping

File: mcpgateway/main.py [hunk 6]
```diff
@@ -1453,8 +1457,29 @@ async def plugin_violation_exception_handler(_request: Request, exc: PluginViola
...
     json_rpc_error = PydanticJSONRPCError(code=status_code, message="Plugin Violation: " + message, data=violation_details)
-    return ORJSONResponse(status_code=200, content={"error": json_rpc_error.model_dump()})
+
+    # Collect HTTP headers from violation if present
+    headers = exc.violation.http_headers if exc.violation and exc.violation.http_headers else None
+
+    response = ORJSONResponse(status_code=http_status, content={"error": json_rpc_error.model_dump()})
+    if headers:
+        response.headers.update(headers)
+    return response
```
Comment (gcgoncalves) [inline mcpgateway/main.py:1547]: The http_headers dict comes directly from plugin code with no validation of header names or values. This allows any plugin (including third-party ones) to:

- Overwrite security headers already set by the existing security middleware (Content-Security-Policy, X-Frame-Options, X-Content-Type-Options, etc.). The exception handler runs after the
middleware chain so response.headers.update() wins.
- Inject CRLF sequences in values — though modern ASGI frameworks often sanitise these, the defence-in-depth approach of not sanitising at all is risky.
- Set Content-Type: text/html on a response whose body happens to contain attacker-controlled content, enabling XSS via the error message field.

Header names and values should be validated against an allowlist (e.g. only X-RateLimit-*, Retry-After, and similar policy headers), and any attempt to override security-sensitive
header names should be logged and dropped.

File: plugins/rate_limiter/rate_limiter.py [hunk 7]
```diff
@@ -87,31 +87,129 @@ class _Window:
...
+    _, limit, reset_ts, meta = most_restrictive
+    remaining = meta.get("remaining", 0)
+    retry_after = meta.get("reset_in", 0)
+
+    # Aggregate metadata from all dimensions
+    aggregated_meta = {
+        "limited": True,
+        "remaining": remaining,
+        "reset_in": retry_after,
+        "dimensions": [m for _, _, _, m in allowed_dims],
```
Comment (gcgoncalves) [inline plugins/rate_limiter/rate_limiter.py]: Should this be a dict with violated/allowed like the one above, for typing consistency?

Comment (ja8zyjits) [reply plugins/rate_limiter/rate_limiter.py]: Updated.

File: mcpgateway/main.py (hunk 6 above)
Comment (ja8zyjits) [reply mcpgateway/main.py:1547]: Thanks for raising these points — they’re very helpful. For the current work, my focus was specifically on improving the rate‑limiting functionality. As part of that, I’ve already implemented strict RFC 9110‑compliant header sanitization and ensured that CRLF and other forbidden control characters are fully rejected as a defence‑in‑depth measure.

The concerns you’ve highlighted around plugins being able to override security‑sensitive headers, set unsafe Content-Type values, or otherwise influence response metadata are broader ecosystem‑level issues. Addressing them properly will require a more comprehensive approach, including decisions around plugin governance, allowlists/denylists for header manipulation, and how we want to handle third‑party plugin contributions in the future. Currently we own the plugins and hence we have full control over them.

Since this goes beyond the scope of the rate‑limiting fix, I suggest we track these recommendations as a separate task so we can give them the design time they deserve without delaying the current improvement.
Happy to help open that follow‑up issue if that works for the team.

File: mcpgateway/plugins/framework/constants.py (hunk 5 above)
Comment (ja8zyjits) [reply mcpgateway/plugins/framework/constants.py]: Done

**Comment by ja8zyjits** [comment]
Thanks @msureshkumar88 for the suggestions, I have updated the code accordingly.

File: mcpgateway/main.py (hunk 6 above)
Comment (gcgoncalves) [reply mcpgateway/main.py:1547]: Great points, agreed!

**Review by msureshkumar88 — APPROVED** [review]
## ✅ PR #3183 — Approved

**2668 bug follow up rate limiter incorrect http status missing headers and feature gaps refs 2397** by @ja8zyjits

This PR successfully enables plugins to return proper HTTP status codes (429 for rate limiting) and standard headers (X-RateLimit-*, Retry-After), addressing critical HTTP semantics gaps. The implementation is well-designed with comprehensive test coverage (1,500+ lines) and maintains full backward compatibility.

**Security**: 🟢 No critical issues found  
**Issue Coverage**: 3/3 high-priority requirements met (2 medium-priority deferred)  
**Quality**: All checks passed

## Final PR Review Status Summary

### Security Findings: ✅ BOTH COMPLETELY FIXED

Both security findings from the original review have been **fully implemented and comprehensively tested** in the feature branch.

---

### Finding #1: HTTP Status Code Validation - ✅ FIXED & TESTED

**Implementation**: [`mcpgateway/plugins/framework/constants.py:94-137`](mcp-context-forge/mcpgateway/plugins/framework/constants.py:94-137)
- Added `VALID_HTTP_STATUS_CODES` dictionary with all RFC 9110 compliant codes
- Validation at [`main.py:1516`](mcp-context-forge/mcpgateway/main.py:1516)

**Test Coverage**: [`tests/unit/mcpgateway/plugins/framework/test_constants.py`](mcp-context-forge/tests/unit/mcpgateway/plugins/framework/test_constants.py)
- 226 lines of comprehensive tests
- Validates all status codes, ranges, and edge cases

---

### Finding #2: HTTP Header Injection Risk - ✅ FIXED & TESTED

**Implementation**: [`mcpgateway/main.py:1410-1452`](mcp-context-forge/mcpgateway/main.py:1410-1452)
- Added `_validate_http_headers()` function with RFC 9110 compliance
- Validates header names, rejects CRLF, filters CTL characters
- Integrated at lines 1534-1536

**Test Coverage**: [`tests/unit/mcpgateway/test_main_helpers.py:187-306`](mcp-context-forge/tests/unit/mcpgateway/test_main_helpers.py:187-306)
- **NEW FILE**: 122 lines dedicated to header validation testing
- Tests cover:
  - ✅ Valid headers (lines 187-195)
  - ✅ Invalid header names (lines 198-217)
  - ✅ CRLF injection attempts (lines 220-244)
  - ✅ CTL character filtering (lines 247-289)
  - ✅ Edge cases: empty dict, all invalid (lines 292-306)

---

### Code Quality Assessment

**Changes Summary** (compared to base branch):
- `mcpgateway/main.py`: +79 lines
- `mcpgateway/plugins/framework/constants.py`: +92 lines
- `mcpgateway/plugins/framework/models.py`: +6 lines
- `plugins/rate_limiter/rate_limiter.py`: +192 lines
- **Tests**: +1,600 lines across 5 test files

**Quality Metrics**:
- ✅ **Type Safety**: All new code properly typed
- ✅ **Documentation**: Comprehensive docstrings with RFC references
- ✅ **Test Coverage**: Excellent (>95% for new code)
- ✅ **Security**: Both findings fixed with defense-in-depth
- ✅ **Standards Compliance**: RFC 9110 (newer than RFC 7231)
- ✅ **Backward Compatibility**: Fully maintained


---

### Final Recommendation

✅ **APPROVE FOR MERGE**

The PR is **production-ready** with:
- All high-priority requirements met (3/3 from issue #2668)
- Both security findings completely fixed and tested
- Excellent code quality and test coverage
- No blocking issues identified

**Comment by jonpspri** [comment]
### Review follow-up: daily rate unit and IP-based limiting

This PR now also includes support for the **daily rate unit** (`/d`, `/day`) in `_parse_rate()`, addressing medium-priority item #5 from #2668.

The remaining item — **IP-based rate limiting for anonymous requests** — requires adding `client_ip` to `GlobalContext` (a plugin framework change) and has been filed separately as #3349.

With this addition, 4 of 5 acceptance criteria from #2668 are addressed. Consider changing `Fixes #2668` to `Refs #2668` in the PR description so the issue stays open until #3349 is resolved, or close #2668 and track the remaining work via #3349.

**Comment by ja8zyjits** [comment]
This branch has features pushed into it, which increases the scope of the PR.

Closing this and a new PR will be created with the original scope of the PR.

**Comment by ja8zyjits** [comment]
New clean PR created https://github.com/IBM/mcp-context-forge/pull/3383
